### PR TITLE
ipn/ipnlocal: transform default routes into "all but LAN" routes.

### LINKED ISSUE #1177: Default routes should block LAN access by default
Came up while designing/implementing default routing: an interesting use case is being able to share an "exit node" with friends and family. However, that feels much safer to do if that sharing doesn't also grant access to your LAN. In other words, "share my internet" should mean "share access to the internet", not "share access to whatever my subnet router happens to have access to".

In practice, that would mean that enabling default routing on a node would offer access to 0.0.0.0/0 and ::/0, but explicitly block access to:
 - RFC1918 and RFC6598 IPv4 space (LAN and CGNAT)
 - IPv6 ULA space
 - The IP space of directly-connected ethernet segments (e.g. the IPv6 /64 assigned to the LAN by the local internet router)

This would mean that, if I share my internet connection with you, you can access e.g. google.com through my node, but you can't load 192.168.0.1 and get to my router's admin panel.

This is a default only, overridable by explicitly advertising the above-filtered routes in addition to default routes.

Part of #102.

### PR DESCRIPTION
**Pull request by danderson**
Fixed #1177.

Signed-off-by: David Anderson <danderson@tailscale.com>

### CODE REVIEW

**Comment by danderson** [comment]
Some questions/thoughts.

For IPv6, we kinda have to dynamically detect the LAN and filter it out. Should we also keep doing it for IPv4, in addition to the static RFC1928 ranges? On servers in clouds like DO, it has the weird effect of preventing access to other VMs that happen to be on the same /24 as you, even though they're on all-public IPs.

@apenwarr's past argument was that we should be secure by default in that case, and in practice I think we'll probably be okay for a while until someone gets unlucky. But IPv4 does afford us the option of saying "if you meant for it to be a LAN, you should have put it in RFC1918" and call it a day.

Thoughts?

File: net/interfaces/interfaces.go [hunk 1]
```diff
@@ -136,7 +136,7 @@ func (i Interface) IsLoopback() bool { return isLoopback(i.Interface) }
 func (i Interface) IsUp() bool       { return isUp(i.Interface) }
 
 // ForeachInterfaceAddress calls fn for each interface's address on the machine.
```
Comment (bradfitz) [inline net/interfaces/interfaces.go:138]: clarify that these are not IPPrefixes in Masked (https://pkg.go.dev/inet.af/netaddr#IPPrefix.Masked) form and that the IP is the actual specific IP within that prefix?

File: net/interfaces/interfaces.go [hunk 2]
```diff
@@ -404,9 +406,9 @@ var (
 
 // anyInterestingIP reports ips contains any IP that matches
```
Comment (bradfitz) [inline net/interfaces/interfaces.go:407]: reports whether pfxs

**Review by bradfitz — COMMENTED** [review]
Tests?

**Comment by danderson** [comment]
Added test and fixed feedback, PTAL.

File: ipn/ipnlocal/local.go [hunk 3]
```diff
@@ -632,6 +649,42 @@ func (b *LocalBackend) updateFilter(netMap *netmap.NetworkMap, prefs *ipn.Prefs)
...
+	netaddr.MustParseIPPrefix("10.0.0.0/8"),
+	// Tailscale IPv4 range
+	tsaddr.CGNATRange(),
+	// IPv6 Link-local addresses
+	netaddr.MustParseIPPrefix("fe80::/10"),
+	// Tailscale IPv6 range
+	tsaddr.TailscaleULARange(),
+}
+
+// shrinkDefaultRoute returns an IPSet representing the IPs route
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:665]: "the IPs route"? typo/word-o?

File: net/interfaces/interfaces.go [hunk 4]
```diff
@@ -451,11 +457,11 @@ var (
 	v6Global1     = mustCIDR("2000::/3")
 )
 
-// anyInterestingIP reports ips contains any IP that matches
+// anyInterestingIP reports whetehr pfxs contains any IP that matches
```
Comment (bradfitz) [inline net/interfaces/interfaces.go:460]: typo: whether

File: net/interfaces/interfaces.go [hunk 5]
```diff
@@ -189,7 +193,7 @@ func ForeachInterface(fn func(Interface, []netaddr.IP)) error {
 // routing table, and other network configuration.
 // For now it's pretty basic.
 type State struct {
-	InterfaceIPs map[string][]netaddr.IP
+	InterfaceIPs map[string][]netaddr.IPPrefix
```
Comment (bradfitz) [inline net/interfaces/interfaces.go:196]: docs about IP vs mask here too?

**Review by bradfitz — APPROVED** [review]

**Pull request merged**